### ISSUE TITLE
Potential for data corruption as writes are not followed by fsync?

### ISSUE THREAD

**Issue by ioquatix** [libgit2/libgit2#4123]
I am interested in the durability of `libgit2` and found the following note:

    do_fsync whether to do an fsync() after writing (currently ignored)

https://github.com/libgit2/libgit2/blob/89c332e41b12a72d89de40d63bc568c56a2c336a/include/git2/odb_backend.h#L42

### Reproduction steps

Write an object to the odb, update a ref, and suffer a power failure.

### Expected behavior

The ref should point to a valid set of objects.

### Actual behavior

The ref points to missing or partially written objects.

### Version of libgit2 (release number or SHA1)

Current.

It seems like fsync should be the default for reliable and durable behaviour. But, perhaps I'm wrong.

It's not enough to call close, you should call fsync before calling close.

> A successful close does not guarantee that the data has been successfully saved to disk, as the kernel defers writes. It is not common for a file system to flush the buffers when the stream is closed. If you need to be sure that the data is physically stored use fsync(2).

I may be wrong, but it would be good to understand whether the git repository could be in an inconsistent state after a power loss and if this is by design or not.

**Comment by ethomson** [comment]
`git` itself rarely does `fsync`s, but we are introducing an option to enable `fsync` for all ref and object writes:  #4030

**Comment by ioquatix** [comment]
Thanks for this. I made https://github.com/ioquatix/relaxo which is a database made on top of `libgit2`. I wanted to understand more about the durability guarantees of `libgit2`. If anything once this PR is merged, it would be great to add some documentation about the durability of `libgit2`. I'd also be happy to do some testing (e.g. try to create corruption).

**Comment by ethomson** [comment]
If you were interested in starting such a document - or collaborating on something - then I think that would be swell.  Even if you had a series of questions / investigations to start with, that would be nice, so that we can make sure that we address the particular concerns that you're interested in speaking to.

**Comment by ioquatix** [comment]
Okay, well, there was a pretty crazy discussion here: https://www.reddit.com/r/programming/comments/5trx9y/github_ioquatixrelaxo_a_transactional_database/

I ended up asking the following question on SO according to the issue policy of libgit2: http://stackoverflow.com/questions/42205268/is-libgit2-safe-during-power-loss/42205341?noredirect=1#comment71572562_42205341

The main questions I'd have would really be what operations are guaranteed to be durable? e.g. append object, update ref, garbage collection, pack file generation, etc.

It could be nice to have a table showing a list of operations and the durability status.

I would be happy to do some benchmarks of real world data with fsync on and off once it's implemented.

For me, the goal is durability, and it makes sense that `libgit2` is durable by default IMHO. It's almost a selling point (could mention it next to the word reentrant :)

**Comment by pureconfig** [comment]
@ioquatix even with fsync, you only guarantee completion of certain steps towards a commit. A power failure can definitely leave you with  dangling references (best case, in which case a repair might be possible)

**Comment by ethomson** [comment]
Assuming that your controller and disk aren't lying to you about `fsync` completion - say because you have a shiny disk controller - how could you end up with a dangling reference?

**Comment by pureconfig** [comment]
@ethomson as in, fsync guarantees individual flushes, not distinct fsync's individually. I assume libgit2 does several fsync-required operations in a commit? (don't forget the need to fsync parent directories)

**Comment by ethomson** [comment]
What do you mean by "commit"?  Are you talking about the general concept of a "commit" whereby you make some set of changes permanent?  Or are you talking about the concept of a "git commit"?

If the latter, what function or set of functions are we talking about exactly?

**Comment by pureconfig** [comment]
> If the latter, what function or set of functions are we talking about exactly?

I believe the topic of this issue is atomicity, not just fsync succeeding. There's nothing ACID about odb_loose.c even with fsync.

**Comment by ioquatix** [comment]
@pureconfig The file is written to a transient location, and moved into place in a single atomic operation. If it crashes during or writing the header, it's garbage and will never be used.

**Comment by ioquatix** [comment]
@pureconfig I think we can all agree that if the underlying hardware is not respecting fsync, there is very little you can do to guarantee anything about the state of the file system during a power loss event, so that situation is really out of our hands w.r.t. how libgit2 works.

**Comment by ioquatix** [comment]
> I assume libgit2 does several fsync-required operations in a commit

To update a ref to the new commit/tree, only a single file is modified.

To put it another way, the commit is made up by writing into the object data store. Any crash during this time doesn't break the object store (or if it does, it's a bug), nor does it affect the visible head of the git repository. Once all those writes are flushed to disk, a single write/flush is used to update the ref.

**Comment by pureconfig** [comment]
> To update a ref to the new commit/tree, only a single file is modified.

But according to the code, there's no system for validating the the "single file" is fully written - I pointed to some code above where a header may end up without the subsequent data record (a write call may end up being flushed by the OS)

**Comment by ioquatix** [comment]
> But according to the code, there's no system for validating the the "single file" is fully written.

Perhaps it's best we reason about a specific timeline.

Here is an strace of an update to the object database:

```strace
open("$repo/objects/tmp_object_git2_AwFNTE", O_RDWR|O_CREAT|O_EXCL, 0600) = 7
chmod("$repo/objects/tmp_object_git2_AwFNTE", 0444) = 0
lstat("$repo/objects/09", 0x7ffed1d8bac0) = -1 ENOENT (No such file or directory)
mkdir("$repo/objects/09", 0777) = 0
write(7, "x\1\255\316A\n\3020\20Fa\327=\305\354\205\322&\245\216 \"x\4\27\256'\351\37\f4\t"..., 146) = 146
close(7)                                = 0
link("$repo/objects/tmp_object_git2_AwFNTE", "$repo/objects/09/8962f12cdfbca041196ddf4496258b60863941") = 0
unlink("$repo/objects/tmp_object_git2_AwFNTE") = 0
```

So, excepting the lack of `fsync`, how does `$repo/objects/09/8962f12cdfbca041196ddf4496258b60863941` end up being in an inconsistent state?

**Comment by pureconfig** [comment]
@ioquatix Because the write may end up as several writes.

**Comment by ioquatix** [comment]
Okay, so let's assume that:

```strace
open("$repo/objects/tmp_object_git2_AwFNTE", O_RDWR|O_CREAT|O_EXCL, 0600) = 7
chmod("$repo/objects/tmp_object_git2_AwFNTE", 0444) = 0
lstat("$repo/objects/09", 0x7ffed1d8bac0) = -1 ENOENT (No such file or directory)
mkdir("$repo/objects/09", 0777) = 0
write(7, "x\1\255\316A\n\3020\20Fa\327=\305\354\205\322&\245\216 \"x\4\27\256'\351\37\f4\t"..., 146) = 146
write(7, "x\1\255\316A\n\3020\20Fa\327=\305\354\205\322&\245\216 \"x\4\27\256'\351\37\f4\t"..., 146) = 146
write(7, "x\1\255\316A\n\3020\20Fa\327=\305\354\205\322&\245\216 \"x\4\27\256'\351\37\f4\t"..., 146) = 146
write(7, "x\1\255\316A\n\3020\20Fa\327=\305\354\205\322&\245\216 \"x\4\27\256'\351\37\f4\t"..., 146) = 146
close(7)                                = 0
link("$repo/objects/tmp_object_git2_AwFNTE", "$repo/objects/09/8962f12cdfbca041196ddf4496258b60863941") = 0
unlink("$repo/objects/tmp_object_git2_AwFNTE") = 0
```

So, excepting the lack of fsync, how does $repo/objects/09/8962f12cdfbca041196ddf4496258b60863941 end up being in an inconsistent state?

**Comment by pureconfig** [comment]
You example is not relevant. What's relevant is the write(<size>, header) followed by power loss/crash/whatever.

**Comment by ioquatix** [comment]
Okay so you are proposing the following sequence of events:

```strace
open("$repo/objects/tmp_object_git2_AwFNTE", O_RDWR|O_CREAT|O_EXCL, 0600) = 7
chmod("$repo/objects/tmp_object_git2_AwFNTE", 0444) = 0
lstat("$repo/objects/09", 0x7ffed1d8bac0) = -1 ENOENT (No such file or directory)
mkdir("$repo/objects/09", 0777) = 0
write(7, "x\1\255\316A\n\3020\20Fa\327=\305\354\205\322&\245\216 \"x\4\27\256'\351\37\f4\t"..., 146) = 146
(crash)
```

Then, all that's potentially left on disk is the temporary object file `$repo/objects/tmp_object_git2_AwFNTE`. This won't affect anything or have any adverse effect on the repository.

**Comment by pureconfig** [comment]
No, I am not proposing that sequence of events. Look at the code I pointed to.

**Comment by ioquatix** [comment]
https://github.com/libgit2/libgit2/blob/a59545de5e57769f858ed2657b9fb850675f227f/src/odb_loose.c#L906-L907

You are referring to those two lines right?

**Comment by ioquatix** [comment]
https://github.com/libgit2/libgit2/blob/a59545de5e57769f858ed2657b9fb850675f227f/src/odb_loose.c#L906-L907

@pureconfig You are referring to those two lines right?
